File: python/N으로표현.py
def dp(n, number, x, c, a):
    if x == number:
        return c
    elif c > 8:
        return 9
    
    a = min(a, dp(n, number, x+n, c+1, a))
    a = min(a, dp(n, number, x-n, c+1, a))
    a = min(a, dp(n, number, x*n, c+1, a))
    a = min(a, dp(n, number, x/n, c+1, a))
    a = min(a, dp(n, number, int(str(n)+str(n)), c+2, a))
    a = min(a, dp(n, number, x+int(str(n)+str(n)), c+2, a))
    a = min(a, dp(n, number, x-int(str(n)+str(n)), c+2, a))
    a = min(a, dp(n, number, x*int(str(n)+str(n)), c+2, a))
    a = min(a, dp(n, number, x/int(str(n)+str(n)), c+2, a))
    a = min(a, dp(n, number, int(str(n)+str(n)+str(n)), c+3, a))
    a = min(a, dp(n, number, x+int(str(n)+str(n)+str(n)), c+3, a))
    a = min(a, dp(n, number, x-int(str(n)+str(n)+str(n)), c+3, a))
    a = min(a, dp(n, number, x*int(str(n)+str(n)+str(n)), c+3, a))
    a = min(a, dp(n, number, x/int(str(n)+str(n)+str(n)), c+3, a))
    a = min(a, dp(n, number, int(str(n)+str(n)+str(n)+str(n)), c+4, a))
    a = min(a, dp(n, number, x+int(str(n)+str(n)+str(n)+str(n)), c+4, a))
    a = min(a, dp(n, number, x-int(str(n)+str(n)+str(n)+str(n)), c+4, a))
    a = min(a, dp(n, number, x*int(str(n)+str(n)+str(n)+str(n)), c+4, a))
    a = min(a, dp(n, number, x/int(str(n)+str(n)+str(n)+str(n)), c+4, a))
    
    return a
    

def solution(n, number):
    answer = 0
    
    answer = dp(n, number, 0, 0, 9)

    if answer > 8:
        return -1
    return answer

print(solution(5, 110))

File: JavaScript/N으로표현.js
function solution(N, number) {
    const answer = calc(N, number, N, 1);
    return answer < 9 ? answer : -1;
}

function calc(N, number, current, count) {
    if (current === number || count > 8) return count;
    
    if (Number.isInteger(current)) {
        let c = 9, n = '';
        for (let i = 1; i < 6; i++) {
            n += String(N);
            c = Math.min(c, calc(N, number, Number(current + Number(n)), count + i));
            c = Math.min(c, calc(N, number, Number(current - Number(n)), count + i));
            c = Math.min(c, calc(N, number, Number(Number(n) - current), count + i));
            c = Math.min(c, calc(N, number, Number(current * Number(n)), count + i));
            c = Math.min(c, calc(N, number, Number(current / Number(n)), count + i));
            c = Math.min(c, calc(N, number, Number(Number(n) / current), count + i));
            c = Math.min(c, calc(N, number, Number(current + n), count + i));
        }
        return c;
    } else return 9;
}

console.log(solution(5, 12));
console.log(solution(2, 11));
console.log(solution(8, 5800));
console.log(solution(5, 31168));